### PR TITLE
Fixes for ghc-8.0 and persistent 2.5

### PR DESCRIPTION
**Pull request by erikd**

### CODE REVIEW

**Comment by erikd** [comment]
This failed in travis-ci because persistent 2.5 has not yet been formally released.

**Comment by snoyberg** [comment]
@gregwebs do you want to merge this after releasing persistent?

**Comment by gregwebs** [comment]
2.5 was released yesterday. I just clicked the redo button on Travis

**Comment by snoyberg** [comment]
Cool, thanks!

**Comment by erikd** [comment]
Still not building. I'll take a look some time today.

**Comment by erikd** [comment]
Building yesod with ghc-8.0 will also depend on https://github.com/prowdsponsor/nonce/issues/3

**Comment by felixonmars** [comment]
yesod-form still fails to build after the patch:

```
Yesod/Form/Fields.hs:660:22:
    Could not deduce (persistent-2.5:Database.Persist.Class.PersistStore.BaseBackend
                        (YesodPersistBackend site)
                      ~ YesodPersistBackend site)
    from the context (YesodPersist site,
                      PersistEntity a,
                      PersistQuery (PersistEntityBackend a),
                      PathPiece (Key a),
                      RenderMessage site msg,
                      YesodPersistBackend site ~ PersistEntityBackend a)
      bound by the type signature for
                 optionsPersist :: (YesodPersist site, PersistEntity a,
                                    PersistQuery (PersistEntityBackend a), PathPiece (Key a),
                                    RenderMessage site msg,
                                    YesodPersistBackend site ~ PersistEntityBackend a) =>
                                   [Filter a]
                                   -> [SelectOpt a]
                                   -> (a -> msg)
                                   -> HandlerT site IO (OptionList (Entity a))
      at Yesod/Form/Fields.hs:(648,19)-(657,58)
    Relevant bindings include
      optionsPersist :: [Filter a]
                        -> [SelectOpt a]
                        -> (a -> msg)
                        -> HandlerT site IO (OptionList (Entity a))
        (bound at Yesod/Form/Fields.hs:658:1)
    In the second argument of `($)', namely `selectList filts ords'
    In a stmt of a 'do' block: pairs <- runDB $ selectList filts ords
    In the second argument of `($)', namely
      `do { mr <- getMessageRender;
            pairs <- runDB $ selectList filts ords;
            return
            $ map
                (\ (Entity key value)
                   -> Option
                        {optionDisplay = mr (toDisplay value),
                         optionInternalValue = Entity key value,
                         optionExternalValue = toPathPiece key})
                pairs }'

Yesod/Form/Fields.hs:686:22:
    Could not deduce (persistent-2.5:Database.Persist.Class.PersistStore.BaseBackend
                        (YesodPersistBackend site)
                      ~ YesodPersistBackend site)
    from the context (YesodPersist site,
                      PersistEntity a,
                      PersistQuery (PersistEntityBackend a),
                      PathPiece (Key a),
                      RenderMessage site msg,
                      YesodPersistBackend site ~ PersistEntityBackend a)
      bound by the type signature for
                 optionsPersistKey :: (YesodPersist site, PersistEntity a,
                                       PersistQuery (PersistEntityBackend a), PathPiece (Key a),
                                       RenderMessage site msg,
                                       YesodPersistBackend site ~ PersistEntityBackend a) =>
                                      [Filter a]
                                      -> [SelectOpt a]
                                      -> (a -> msg)
                                      -> HandlerT site IO (OptionList (Key a))
      at Yesod/Form/Fields.hs:(672,6)-(682,42)
    Relevant bindings include
      optionsPersistKey :: [Filter a]
                           -> [SelectOpt a]
                           -> (a -> msg)
                           -> HandlerT site IO (OptionList (Key a))
        (bound at Yesod/Form/Fields.hs:684:1)
    In the second argument of `($)', namely `selectList filts ords'
    In a stmt of a 'do' block: pairs <- runDB $ selectList filts ords
    In the second argument of `($)', namely
      `do { mr <- getMessageRender;
            pairs <- runDB $ selectList filts ords;
            return
            $ map
                (\ (Entity key value)
                   -> Option
                        {optionDisplay = mr (toDisplay value), optionInternalValue = key,
                         optionExternalValue = toPathPiece key})
                pairs }'
```

**Comment by erikd** [comment]
I have a fix for `yesod-form` and now working on `yesod-auth`.

**Comment by erikd** [comment]
The travis-ci failure seems to be stack related. I'm still building with cabal.

**Comment by snoyberg** [comment]
This is a real error, since the lower bounds are too strict. We need support for both old and new persistent via CPP.

**Comment by erikd** [comment]
@snoyberg You mean in `yesod-persistent` we need to support both old and new?

**Comment by snoyberg** [comment]
Yes

**Comment by erikd** [comment]
Ok, working on it.

**Comment by snoyberg** [comment]
Thanks!

**Comment by erikd** [comment]
@gregwebs @snoyberg If we want yesod to support both old persisitent (version < 2.5) and new persistent (version == 2.5.*) as well as supporting ghc-8.0, then we need to release a version of persistent-template that has my ghc-8.0 fixes, but without the read/write split.

Happy to do this, but just want to make sure that is what you want.

**Comment by erikd** [comment]
If I checkout the `persistent-template/2.1.6` tag as a branch I can then cherry pick commit 3adc3e6df021 which then builds and tests ok with both ghc-7.10 and ghc-8.0. I'd do a pull request, but I'm not sure how to do a pull request against something that is not a branch.

**Comment by erikd** [comment]
Hopefully this latest set of commits works for persistent < 2.5 as well as persistent == 2.5 as well as working for ghc-7.10 and ghc-8.0.

**Comment by erikd** [comment]
Oh crap, using `<$>` breaks for ghc-7.8.

**Comment by mgsloan** [comment]
Usage of liftM is fine :)

**Comment by erikd** [comment]
Bah! Seem to have messed up a merge somehwere.

**Comment by erikd** [comment]
Nope, commit d99de615547d messed up my WIP patches.

**Comment by erikd** [comment]
Yay! That turned out to be a lot more work than I initially expected.

**Comment by snoyberg** [comment]
Thanks for doing it, will review today

File: yesod-core/Yesod/Routes/TH/RenderRoute.hs [hunk 1]
```diff
@@ -8,26 +8,30 @@ module Yesod.Routes.TH.RenderRoute
...
 import Data.Monoid (mconcat)
 #endif
 
 -- | Generate the constructors of a route data type.
-mkRouteCons :: [ResourceTree Type] -> ([Con], [Dec])
-mkRouteCons =
-    mconcat . map mkRouteCon
+mkRouteCons :: [ResourceTree Type] -> Q ([Con], [Dec])
+mkRouteCons rttypes =
+    mconcat <$> mapM mkRouteCon rttypes
```
Comment (snoyberg) [inline yesod-core/Yesod/Routes/TH/RenderRoute.hs]: Was this change necessary?

File: yesod-core/test/YesodCoreTest/Exceptions.hs [hunk 2]
```diff
@@ -41,7 +41,7 @@ case500 :: IO ()
 case500 = runner $ do
     res <- request defaultRequest
     assertStatus 500 res
-    assertBody "FOOBAR" res
+    assertBodyContains "FOOBAR" res
```
Comment (snoyberg) [inline yesod-core/test/YesodCoreTest/Exceptions.hs]: It's strange that this test case needed to change, what motivated that tweak?

Comment (erikd) [inline yesod-core/test/YesodCoreTest/Exceptions.hs]: In ghc-8.0, `error` includes a call stack and the default `Show` instance for exceptions prints the callstack.

```> ghci -XScopedTypeVariables
GHCi, version 8.0.0.20160411: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/erikd/.ghci
ghci > import Control.Exception
ghci > catch (error "FOOBAR") (\(e :: SomeException) -> print e)
FOOBAR
CallStack (from HasCallStack):
```

File: yesod-core/Yesod/Routes/TH/RenderRoute.hs (hunk 1 above)
Comment (erikd) [inline yesod-core/Yesod/Routes/TH/RenderRoute.hs]: Yes, `mkRouteCon` used to be pure (returning `Dec`), but now it needs to run in the `Q` monad (now returning `Q Dec`) so that `mapM conT` can be used in the `MIN_VERSION_template_haskell(2,11,0)` case.

Comment (erikd) [inline yesod-core/Yesod/Routes/TH/RenderRoute.hs]: Now that I've had my morning coffee, I can give a more coherent explanation :).

As described in the GHC 8.0 mirgation guide ( https://ghc.haskell.org/trac/ghc/wiki/Migration/8.0 ), the `DataD` constructor has changed:

```
    - dataD :: CxtQ -> Name -> [TyVarBndr]               -> [ConQ] -> [Name] -> DecQ
    + dataD :: CxtQ -> Name -> [TyVarBndr] -> Maybe Kind -> [ConQ] -> CxtQ   -> DecQ
```

Specifiacally the last field has changed from `[Name]` to `CxtQ` which is just a type synonym for `Q Cxt`. The existing code for ghc < 8.0 already had value of type `[Name]` and the way to convert value `xs :: [Name]` to `CxtQ` is to use `mapM conT xs`.

Once we need the `Q` monad there, the whole function needs to be in the `Q` monad.

File: yesod-core/test/YesodCoreTest/Exceptions.hs (hunk 2 above)
Comment (snoyberg) [inline yesod-core/test/YesodCoreTest/Exceptions.hs]: I'm still not following, sorry. There's a change here from `assertBody` to `assertBodyContains`, which is the difference between "is the body exactly X" and "does the body contain X." What's the relevance of `error` to that change?

File: yesod-core/Yesod/Routes/TH/RenderRoute.hs (hunk 1 above)
Comment (snoyberg) [inline yesod-core/Yesod/Routes/TH/RenderRoute.hs]: Thanks for the explanation, I wasn't looking at enough of the code.

Comment (erikd) [inline yesod-core/Yesod/Routes/TH/RenderRoute.hs]: :-).

File: yesod-core/test/YesodCoreTest/Exceptions.hs (hunk 2 above)
Comment (erikd) [inline yesod-core/test/YesodCoreTest/Exceptions.hs]: Hmm, i really should not respond to github issues before my morning coffee :).

Without the change from `assertBody` to `assertBodyContains` after building with ghc-8.0, the test fail with:

```
  1) Test.Exceptions 500
       uncaught exception: WaiTestFailure (WaiTestFailure "Expected response body \"FOOBAR\", but received \"FOOBAR\\nCallStack (from HasCallStack):\\n  error, called at test/YesodCoreTest/Exceptions.hs:25:12 in main:YesodCoreTest.Exceptions\"")
```

which points directly at line 25 which is:

```
getRootR = error "FOOBAR" >> return ()
```

So, `error` is relevant, because `error` now has a `CallStack`.

Comment (snoyberg) [inline yesod-core/test/YesodCoreTest/Exceptions.hs]: Ah, OK, that makes sense. I'm on board with that change then. Merging, thank you for all the work on this!
